src/realm/ucx/ucc_comm.cc

```diff
@@ -147,50 +226,65 @@ namespace Realm {
     ucc_status_t UCCComm::init()
     {
       ucc_status_t status{UCC_OK};
+      unsigned major = 0, minor = 0, release = 0;
+
+      if(!resolve_fnptrs()) {
+        status = UCC_ERR_NOT_FOUND;
+        goto Done;
+      }
+
+      UCC_FUNC(ucc_get_version)(&major, &minor, &release);
+
+      if(UCC_API_VERSION > UCC_VERSION(major, minor)) {
+        log_ucc.error("UCC version mismatch: compiled with %d, found %d", UCC_API_VERSION,
+                      UCC_VERSION(major, minor));
+        goto Done;
+      }
 
-      status = init_lib();
+      status = init_lib(lib);
       if(UCC_OK != status) {
-        return status;
+        goto Done;
       }
 
-      status = create_context();
+      status = create_context(lib, oob_comm, context);
       if(UCC_OK != status) {
-        return status;
+        goto Done;
       }
 
-      status = create_team();
+      status = create_team(context, oob_comm, team);
       if(UCC_OK != status) {
-        return status;
+        goto Done;
       }
-      return status;
-    }
 
-    inline void UCCComm::ucc_check(const ucc_status_t &status)
-    {
-      if(status < 0) {
-        std::cerr << "UCC: Failed " << std::string(ucc_status_string(status))
-                  << std::endl;
-        assert(0);
+    Done:
+      if(status != UCC_OK) {
+        UCC_Finalize();
       }
+
+      return status;
     }
 
     ucc_status_t UCCComm::ucc_collective(ucc_coll_args_t &coll_args, ucc_coll_req_h &req)
     {
       ucc_status_t status;
-      status = ucc_collective_init(&coll_args, &req, team);
-      ucc_check(status);
-
-      status = ucc_collective_post(req);
-      ucc_check(status);
+      status = UCC_FUNC(ucc_collective_init)(&coll_args, &req, team);
+      if(status != UCC_OK) {
+        return status;
+      }
 
-      while(UCC_OK != (status = ucc_collective_test(req))) {
-        ucc_check(status);
-        status = ucc_context_progress(context);
-        ucc_check(status);
+      status = UCC_FUNC(ucc_collective_post)(req);
+      if(status != UCC_OK) {
+        goto Done;
       }
 
-      status = ucc_collective_finalize(req);
-      ucc_check(status);
+      status = wait(context, [&req]() { return req->status; });
```

Comment:
NIT: it is better to use `ucc_collective_test` even though it is the same as req->status, we do not need to know the implementation of a request.